client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +787,139 @@ where
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
+}
+
+impl<Block> EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	fn new() -> Self {
+		EnactmentState { recent_best_block: None, recent_finalized_block: None }
+	}
+
+	/// This function returns true and the tree_route if blocks enact/retract process (implemented
+	/// in handle_enactment function) should be performed based on:
+	/// - recent_finalized_block
+	/// - recent_best_block
+	/// - newly provided block in event
+	/// tree_route contains blocks to be enacted/retracted.
+	///
+	/// if enactment process is not needed (false, None) is returned
+	/// error message is returned when computing tree_route fails
+	pub fn update_and_check_if_new_enactment_is_valid<PoolApi: graph::ChainApi<Block = Block>>(
+		&mut self,
+		api: &PoolApi,
+		event: &ChainEvent<Block>,
+	) -> Result<(bool, Option<TreeRoute<Block>>), String> {
+		let (new_hash, finalized) = match event {
+			ChainEvent::NewBestBlock { hash, .. } => (*hash, false),
+			ChainEvent::Finalized { hash, .. } => (*hash, true),
+		};
+
+		let best_block = match self.recent_best_block {
+			Some(recent_best_hash) => recent_best_hash,
+			None => {
+				self.recent_best_block = Some(new_hash);
+				if finalized {
+					self.recent_finalized_block = Some(new_hash);
+				}
+
+				return Ok((true, None))
+			},
+		};
+
+		// compute actual tree route from best_block to notified block, and use it instead of
+		// tree_route provided with event
+		let tree_route = match api.tree_route(best_block, new_hash) {
+			Ok(tree_route) => tree_route,
+			Err(e) =>
+				return Err(format!(
+					"Error [{e}] occured while computing tree_route from {new_hash:?} to \
+									previously finalized: {best_block:?}"
+				)),
+		};
+
+		log::trace!(
+			target: "txpool", 
+			"resolve hash:{:?} finalized:{:?} tree_route:{:?} best_block:{:?} finalized_block:{:?}",
+			new_hash, finalized, tree_route, best_block, self.recent_finalized_block
+		);
+
+		if let Some(finalized_block) = self.recent_finalized_block {
+			// block was already finalized
+			if finalized_block == new_hash {
+				log::trace!(target:"txpool", "handle_enactment: block already finalized");
+				return Ok((false, None))
+			}
+
+			// check if recently finalized block is on retracted path...
+			if tree_route.retracted().iter().any(|x| x.hash == finalized_block) {
+				log::warn!(
+					target: "txpool",
+					"Recently finalized block {} would be retracted by Finalized event {}",
+					finalized_block, new_hash
+				);
```

Comment:
This should not be a warn, a debug at maximum. 

Because as @andresilva already said, it could be a completely fine order of events that lead to this.

Reply:
[29e31b3](https://github.com/paritytech/substrate/pull/12305/commits/29e31b3d2e360e57ba835d3cf8b88ce5112226f4)